### ISSUE TITLE
OpenMP is being wasted in gomp_barrier_wait_end()?

### ISSUE THREAD

**Issue by RedDwarf69** [byuu/bsnes#289]
Let me start saying I know close to nothing about OpenMP, and even less about its use in bsnes. But I have found this playing Aladdin:

- bsnes master without OpenMP: 95-99% CPU usage
- bsnes master without OpenMP and https://github.com/byuu/bsnes/pull/286/commits/a974c411e4cbe4c7201d678c90771f325323a5e4: 50% CPU usage
- bsnes master with OpenMP: 700% CPU (8 cores), it's not even reaching 60fps
- bsnes master with OpenMP and https://github.com/byuu/bsnes/pull/286/commits/a974c411e4cbe4c7201d678c90771f325323a5e4: 500% CPU

This is what perf has to say about the master with OpenMP (no patch) build:
```
  82.81%  bsnes            libgomp.so.1.0.0               [.] gomp_barrier_wait_end
   6.65%  bsnes            [snd_hda_intel]                [k] pci_azx_readl
   1.43%  bsnes            [kernel.kallsyms]              [k] do_syscall_64
   0.83%  bsnes            libgomp.so.1.0.0               [.] gomp_team_barrier_wait_end
   0.43%  bsnes            [kernel.kallsyms]              [k] syscall_return_via_sysret
   0.41%  bsnes            [kernel.kallsyms]              [k] entry_SYSCALL_64
   0.40%  bsnes            bsnes                          [.] SuperFamicom::PPUfast::Line::renderBackground
   0.38%  bsnes            bsnes                          [.] SuperFamicom::CPU::read
   0.34%  bsnes            bsnes                          [.] SuperFamicom::CPU::step<4u, true>
   0.27%  bsnes            bsnes                          [.] SuperFamicom::PPUfast::Line::render
   0.21%  bsnes            bsnes                          [.] SuperFamicom::CPU::step<4u, false>
   0.16%  bsnes            [kernel.kallsyms]              [k] avc_has_extended_perms
   0.15%  bsnes            libpthread-2.30.so             [.] __pthread_mutex_lock
   0.15%  bsnes            libpthread-2.30.so             [.] __pthread_mutex_unlock_usercnt
   0.15%  bsnes            [snd]                          [k] snd_ctl_find_id
   0.13%  bsnes            bsnes                          [.] SuperFamicom::CPU::step<6u, false>
   0.11%  bsnes            bsnes                          [.] Processor::WDC65816::instruction
   0.11%  bsnes            bsnes                          [.] SuperFamicom::SMP::read
   0.10%  bsnes            [kernel.kallsyms]              [k] read_tsc
   0.09%  bsnes            [kernel.kallsyms]              [k] dw_readl
   0.09%  bsnes            bsnes                          [.] SuperFamicom::PPUfast::Line::renderObject
   0.08%  bsnes            [kernel.kallsyms]              [k] copy_user_enhanced_fast_string
   0.07%  bsnes            bsnes                          [.] SuperFamicom::PPUfast::Line::renderMode7
   0.07%  bsnes            libasound.so.2.0.0             [.] snd_pcm_state
   0.07%  bsnes            [kernel.kallsyms]              [k] __fget
   0.06%  bsnes            bsnes                          [.] nall::function<unsigned char (unsigned int, unsigned char)>::member<SuperFamicom::ReadableMemory>::operator()
```

This is with the ALSA audio backend, syncing to audio with 100ms latency.

So basically in my computer:
- Fedora 31
- Linux 5.3.16
- gcc 9.2.1

It runs better without OpenMP, which seems to be wasting most of my CPU anyway in a spin loop?

My CPU is a "Intel(R) Core(TM) i7-8550U CPU @ 1.80GHz".

**Comment by Screwtapello** [comment]
You can find uses of OpenMP in the bsnes codebase by searching for `#pragma omp`. Right now, there's two:

- `ruby/video/xshm.cpp`, the no-hardware-accelleration video driver, which uses OpenMP to try and accelerate scaling the output image.
- `bsnes/sfc/ppu-fast/line.cpp`, the Fast PPU, which uses OpenMP to accelerate rendering each scanline of the SNES output frame.

The XShm driver isn't the default, so I assume you're not using that.

The fast PPU normally has to draw 240 scanlines per frame, or 480 in interlace mode. However if you enable HD Mode 7, then it has to render however many scanlines you select: 480, 720, 960, etc. Because of the aspect ratio, bumping up the number of scanlines also increases the size of each scanline (at 480p, each scanline is ~640px, but at 960p each scanline is ~1280px).

The way OpenMP works is that you write an ordinary C for loop and decorate it with the `#pragma omp` incantation. The compiler inserts instructions to dispatch each iteration of the loop in parallel to a pool of worker-threads, and wait for them all to complete.

So, it may not be a bad thing that so much time is spent in `gomp_barrier_wait_end`; that could just be the main thread waiting for all the worker threads to complete, which is exactly what it's supposed to do. Alternatively, if you have a lot of cores so the thread-pool has a lot of threads, and you're using not using HD Mode 7, there may not be enough work for them to do, so they spend all their time checking for work instead of doing anything.

At any rate, the real way to benchmark this is *not* to look at CPU usage, but to set the Audio  and Video drivers to None (so bsnes runs as fast as possible) and look at the FPS counter.

**Comment by byuu** [comment]
I've always suspected that OpenMP ran horrifically bad, but not been willing enough to implement my own thread dispatcher to prove it.

On every platform I've tried it on, OpenMP helps increase the framerate, both with XShm and with HD Mode 7 (*especially* as the HD Scale value increases), but I suspect that we could do a lot better than OpenMP with a custom thread dispatcher.

**Comment by byuu** [comment]
Not really anything I can do with this, so closing for now. Feel free to reopen if any plan of action can be devised.
